File: Python/Flatten Binary Tree to Linked List.py
"""
Given a binary tree, flatten it to a linked list in-place.
For example,
Given
         1
        / \
       2   5
      / \   \
     3   4   6
The flattened tree should look like:
   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6
"""

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return nothing, do it in place
    def flatten(self, root):
        stack = []
        node = root
        result = []
        while len(stack) >0 or node is not None:
            if node is not None:
                result.append(node)
                if node.right is not None:
                    stack.append(node.right)
                node = node.left
            else:
                node = stack.pop()
        
        if len(result) > 0:
            root = result.pop(0)
        while len(result)>0:
            root.left = None
            root.right = result.pop(0)
            root = root.right

def flatten(self, root):
        while root:
            if root.left:
                pre = root.left
                while pre.right:
                    pre = pre.right
                pre.right = root.right
                root.right = root.left
                root.left = None
            root = root.right

class Solution:
    # @param root, a tree node
    # @return nothing, do it in place
    def flatten(self, root):
        return self.flattenRecu(root, None)
        
    def flattenRecu(self, root, list_head):
        if root != None:
            list_head = self.flattenRecu(root.right, list_head)
            list_head = self.flattenRecu(root.left, list_head)
            root.right = list_head
            root.left = None
            return root
        else:
            return list_head
        
class Solution2:
    list_head = None
    # @param root, a tree node
    # @return nothing, do it in place
    def flatten(self, root):
        if root != None:
            self.flatten(root.right)
            self.flatten(root.left)
            root.right = self.list_head
            root.left = None
            self.list_head = root
            return root

File: Java/Flatten Binary Tree to Linked List.java
/*
Given a binary tree, flatten it to a linked list in-place.

For example,
Given

         1
        / \
       2   5
      / \   \
     3   4   6
The flattened tree should look like:
   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void flatten(TreeNode root) {
        if(root == null) return;
        
        flatten(root.left);
        flatten(root.right);
        
        if(root.left == null) return;
        
        TreeNode p = root.left;
        while(p.right != null) p = p.right;
        p.right = root.right;
        root.right = root.left;
        root.left = null;
    }
}

public class Solution {
    public void flatten(TreeNode root) {
        flatten(root, null);
    }
    
    private TreeNode flatten(TreeNode root, TreeNode tail){
        if(root == null) return tail;
        
        root.right = flatten(root.left, flatten(root.right, tail));
        root.left = null;
        return root;
    } 
}

public class Solution {
    public void flatten(TreeNode root) {
        if(root == null) return;
        Stack<TreeNode> stack = new Stack<>();
        
        stack.push(root);
        
        while(!stack.empty()){
            TreeNode p = stack.peek();
            stack.pop();
            
            if(p.right != null) stack.push(p.right);
            if(p.left != null) stack.push(p.left);
            
            p.left = null;
            if(!stack.empty()){
                p.right = stack.peek();
            }
        }
    }
}
